app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,419 @@
+package io.appium.settings;
+
+import android.hardware.display.DisplayManager;
+import android.hardware.display.VirtualDisplay;
+import android.media.AudioAttributes;
+import android.media.AudioFormat;
+import android.media.AudioPlaybackCaptureConfiguration;
+import android.media.AudioRecord;
+import android.media.MediaCodec;
+import android.media.MediaCodecInfo;
+import android.media.MediaFormat;
+import android.media.MediaMuxer;
+import android.media.projection.MediaProjection;
+import android.os.Build;
+import android.os.Handler;
+import android.util.DisplayMetrics;
+import android.util.Log;
+import android.view.Surface;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import androidx.annotation.RequiresApi;
+
+public class RecorderThread implements Runnable {
+
+    private static final String TAG = "RecorderThread";
+
+    private MediaProjection mediaProjection;
+    private VirtualDisplay virtualDisplay;
+    private Surface surface;
+    private MediaCodec audioEncoder;
+    private MediaCodec videoEncoder;
+    private AudioRecord audioRecord;
+    private MediaMuxer muxer;
+    private boolean muxerStarted;
+    private boolean startTimestampInitialized;
+    private long startTimestampUs;
+    private Handler handler;
+    private Thread audioRecordThread;
+
+    private int videoTrackIndex = -1;
+    private int audioTrackIndex = -1;
+
+
+    private String outputFilePath;
+    private String videoMime;
+    private int videoWidth;
+    private int videoHeight;
+    private int videoBitrate;
+    private int sampleRate;
+    private Thread recordingThread;
+    private boolean isRotated;
+
+    private volatile boolean stopped = false;
+    private volatile boolean audioStopped;
+    private volatile boolean asyncError = false;
+
+    private MediaCodec.BufferInfo bufferInfo;
+    private long lastAudioTimestampUs;
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private VirtualDisplay.Callback displayCallback = new VirtualDisplay.Callback() {
+        @Override
+        public void onPaused() {
+            super.onPaused();
+            Log.v(TAG, "Display paused");
+        }
+
+        @Override
+        public void onStopped() {
+            super.onStopped();
+            if (!stopped) {
+                asyncError = true;
+            }
+        }
+    };
+
+    public RecorderThread(MediaProjection mediaProjection, String outputFilePath,
+                          int videoWidth, int videoHeight, boolean isRotated) {
+        this.mediaProjection = mediaProjection;
+        handler = new Handler();
+        this.outputFilePath = outputFilePath;
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.sampleRate = RecorderConstant.AUDIO_CODEC_SAMPLE_RATE;
+        this.isRotated = isRotated;
+        videoMime = MediaFormat.MIMETYPE_VIDEO_AVC;
+        bufferInfo = new MediaCodec.BufferInfo();
+        lastAudioTimestampUs = -1;
```

Comment:
should we instead put the default value to where the var is defined?

Reply:
same above

Reply:
done